File: 2019/day13.zig
const std = @import("std");
const utils = @import("utils.zig");

const Tile = enum {
    Empty,
    Wall,
    Block,
    Paddle,
    Ball,
};

const Screen = struct {
    const Self = @This();
    tiles: std.AutoHashMap(utils.Pos, Tile) = undefined,
    score: i64 = undefined,

    fn parse(_out: []i64) Self {
        var s = Self{ .tiles = utils.makeMap(utils.Pos, Tile) };
        var out = _out;
        while (out.len > 0) : (out = out[3..]) {
            var p = utils.Pos{ .x = out[0], .y = out[1] };
            if (p.x == -1 and p.y == 0) {
                s.score = out[2];
            } else {
                _ = s.tiles.put(p, @intToEnum(Tile, @intCast(u3, out[2]))) catch unreachable;
            }
        }
        return s;
    }
};

pub fn main() void {
    var prog = utils.loadProgram("day13_input.txt");

    // part 1
    var screen = Screen.parse(utils.Machine.init(prog).run(0) orelse unreachable);
    var n: usize = 0;
    var mit = screen.tiles.iterator();
    while (mit.next()) |kv| {
        if (kv.value == .Block) n += 1;
    }
    utils.println(n);

    // part 2
    var tilt: i64 = 0;
    var ballX: i64 = undefined;
    var paddleX: i64 = undefined;
    prog[0] = 2;
    var m = utils.Machine.init(prog);
    while (true) {
        screen = Screen.parse(m.run(tilt) orelse break);
        mit = screen.tiles.iterator();
        while (mit.next()) |kv| {
            switch (kv.value) {
                .Paddle => {
                    paddleX = kv.key.x;
                },
                .Ball => {
                    ballX = kv.key.x;
                },
                else => {},
            }
        }
        tilt = utils.sign(i64, ballX - paddleX);
    }
    utils.println(screen.score);
}

File: 2019/day13.go
package main

import (
	"github.com/lukechampine/advent/2019/intcode"
	"github.com/lukechampine/advent/utils"
)

var input = utils.ReadInput("day13_input.txt")
var prog = utils.ExtractInts(input)

func parseTiles(out []int) map[utils.Pos]int {
	tiles := make(map[utils.Pos]int)
	for len(out) > 0 {
		x, y, id := out[0], out[1], out[2]
		out = out[3:]
		tiles[utils.Pos{X: x, Y: y}] = id
	}
	return tiles
}

func main() {
	// part 1
	tiles := parseTiles(intcode.New(prog).Run(0))
	var n int
	for _, id := range tiles {
		if id == 2 {
			n++
		}
	}
	utils.Println(n)

	// part 2
	var tilt int
	var ballX int
	var paddleX int
	prog[0] = 2
	m := intcode.New(prog)
	for !m.Halted {
		tiles = parseTiles(m.Run(tilt))
		for p, id := range tiles {
			switch id {
			case 3:
				paddleX = p.X
			case 4:
				ballX = p.X
			}
		}
		switch {
		case ballX < paddleX:
			tilt = -1
		case ballX == paddleX:
			tilt = 0
		case ballX > paddleX:
			tilt = 1
		}
	}
	utils.Println(tiles[utils.Pos{X: -1, Y: 0}])
}
